tests/integration/test_jenkins.py

```diff
@@ -29,6 +32,94 @@ async def test_jenkins_update_ui_disabled(
+    charm_path = f"/var/lib/juju/agents/unit-{unit_name.replace('/', '-')}/charm"
+
+    # Overwrite requirements.txt
+    with tempfile.NamedTemporaryFile("wt", encoding="utf-8") as temp_requirements_txt, open(
+        "requirements.txt", encoding="utf-8"
+    ) as orig_requirements_txt:
+        requirements = "\n".join((orig_requirements_txt.read(), "freezegun"))
+        temp_requirements_txt.write(requirements)
+        temp_requirements_txt.flush()
+
+        await ops_test.juju(
+            "scp", temp_requirements_txt.name, f"{unit_name}:{charm_path}/requirements.txt"
+        )
+
+        # Install requirements.txt
+        await ops_test.juju(
+            "ssh",
+            unit_name,
+            "/usr/local/bin/pip",
+            "install",
+            "--target",
+            f"{charm_path}/venv",
+            "-r",
+            f"{charm_path}/requirements.txt",
+        )
+
+    # Overwrite charm.py
+    with tempfile.NamedTemporaryFile(mode="wt", encoding="utf-8") as temp_charm_py, open(
+        "src/charm.py", encoding="utf-8"
+    ) as orig_charm_py:
+        shebang_delimeter = "#!/usr/bin/env python3"
+        class_delimeter = "class JenkinsK8SOperatorCharm(CharmBase):"
+        charm_code = orig_charm_py.read()
+        split_code = charm_code.split(shebang_delimeter, 1)
+        split_code.insert(0, f"{shebang_delimeter}\nfrom freezegun import freeze_time\n")
+        charm_code = "".join(split_code)
+        split_code = charm_code.split(class_delimeter, 1)
+        split_code.insert(1, f'freeze_time("{freeze_time}").start()\n{class_delimeter}\n')
+        temp_charm_py.write("".join(split_code))
+        temp_charm_py.flush()
+
+        # unit.scp_to currently doesn't support k8s charms
+        # https://github.com/juju/python-libjuju/issues/885
+        (ret, _, stderr) = await ops_test.juju(
+            "scp", temp_charm_py.name, f"{unit_name}:{charm_path}/src/charm.py"
+        )
+        assert not ret, f"failed to scp modified charm, {stderr}"
+        (ret, _, stderr) = await ops_test.juju(
+            "ssh", unit_name, "/usr/bin/chmod", "755", f"{charm_path}/src/charm.py"
+        )
+        assert not ret, f"failed to chmod modified charm, {stderr}"
+        (ret, _, stderr) = await ops_test.juju(
+            "ssh", unit_name, "/usr/bin/chown", "root:root", f"{charm_path}/src/charm.py"
+        )
+        assert not ret, f"failed to chown modified charm, {stderr}"
+
+    # Wait for model update-status hook
+    await model_app_unit.model.set_config({"update-status-hook-interval": "10s"})
+    time.sleep(15)
+    await model_app_unit.model.wait_for_idle(status="active")
+
+    model_status: client.FullStatus = await model_app_unit.model.get_status()
+    app_status = model_status.applications.get(model_app_unit.app.name)
+    assert app_status, "application status not found."
+
+    assert app_status.workload_version == jenkins_version
+
+    # Reset charm.py & model hook interval
```

Comment:
Should this be done in a fixture? This code won't get executed on a failure

Reply:
Refactored to fixture, thank you for the suggestion!